crates/ra_hir_def/src/body/lower.rs

```diff
@@ -616,12 +618,37 @@ where
 
                 Pat::Record { path, args: fields }
             }
+            ast::Pat::SlicePat(p) => {
+                let mut args = p.args().peekable();
+                let prefix = args
+                    .peeking_take_while(|p| match p {
```

Comment:
This bit mostly does syntactic analysis, so I think we should move it to ra_syntax, extensions.rs specifically.

```rust
struct SlicePatComponents {
    prefix: Vec<Pat>,
    // which name rustc uses for this? rest doesn't seem like a perfect one
    rest: Option<Pat>,
    suffix: Vec<Pat>,
}

impl ast::SlicePat {
    fn components(&self) -> SlicePatCompoents {... }
}
```

Not sure exactly about the precise shape of the type we want here

Reply:
Re. naming, from https://github.com/rust-lang/rust/pull/67712:

> Common to both flavors is they use the token `..`, referred as a *"rest pattern"* in a pattern context.

Reply:
`rustc` uses the naming convention:
```rust
    Slice {
        before / prefix: List<Pat>,
        slice: Option<Pat>,
        after / suffix: List<Pat>,
    },
```
where `before` and `after` are used in HIR and `prefix` and `suffix` are used in HAIR. I would recommend going with `prefix` and `suffix` (as I want to change HIR to that some day).

> Common to both flavors is they use the token `..`, referred as a _"rest pattern"_ in a pattern context.

This refers to the syntactic pattern production `Pat = ... |  Rest:".." ;` and is not specific to slice patterns. The semantic `{binding @}? @ ..` is called a sub-slice. (`syn` also uses `Rest`.)